Add Flower baselines template

jafermarq opened the pull request:
This:
1.  introduces a template for baseline contributors. 
2. edits the top-level readme for the baselines
3. adds a script (`dev/create_baseline.sh`) that users can rely on to clone and rename the teample_baseline

danieljanes commented on baselines/README.md:
````diff
@@ -1,22 +1,31 @@
+2. Navigate to the `baselines/` directory and from there run:
+    ```bash
+    # This will create a new directory with the same structure as `baseline_template`.
+    ./dev/create_baseline.sh <your_baseline_name>
````
```suggestion
    ./dev/create-baseline.sh <your_baseline_name>
```

danieljanes commented on baselines/baseline_template/.gitignore:
```diff
@@ -0,0 +1,2 @@
+*__pycache__/
+*.DS_Store
```
Should this be a full Python `.gitignore`? Or can we remove it because it's redundant with the root `.gitignore`?

danieljanes commented on baselines/baseline_template/EXTENDED_README.md:
```diff
@@ -0,0 +1,52 @@
+
+# Extended Readme
+
+> The baselines are expected to run in a machine running Ubuntu
```
Should we specify a version here?
```suggestion
> The baselines are expected to run in a machine running Ubuntu 22.04
```

danieljanes commented on baselines/baseline_template/EXTENDED_README.md:
````diff
@@ -0,0 +1,52 @@
+2. Navigate to the `baselines/` directory and from there run:
+    ```bash
+    # This will create a new directory with the same structure as this `baseline_template` directory.
+    ./dev/create_baseline.sh <your_baseline_name>
````
```suggestion
    ./dev/create-baseline.sh <your_baseline_name>
```

danieljanes commented on baselines/baseline_template/EXTENDED_README.md:
```diff
@@ -0,0 +1,52 @@
+    *    A `pyproject.toml` detailing the Python environment construction process via [Poetry](https://python-poetry.org/docs/)
+        *    Make sure the variable `name` in `pyproject.toml` is set to the name of the sub-directory containing all your code.
+6. Add your dependencies to the `pyproject.toml` (See below a few examples on how to do it)
+7. Ensure that the Python environment for your basline can be created without errors by simply running: `poetry install`
```
```suggestion
7. Ensure that the Python environment for your baseline can be created without errors by simply running: `poetry install`
```

danieljanes commented on baselines/baseline_template/EXTENDED_README.md:
```diff
@@ -0,0 +1,52 @@
+        *    Make sure the variable `name` in `pyproject.toml` is set to the name of the sub-directory containing all your code.
+6. Add your dependencies to the `pyproject.toml` (See below a few examples on how to do it)
+7. Ensure that the Python environment for your basline can be created without errors by simply running: `poetry install`
+8. Ensure that your baseline runs with default argument by running `poetry run python -m <your_baseline_name>/main.py`. Then, follow the instructions provided in the `README.md` and detail the steps to follow in `Environment Setup` and in `Running the Experiments`.
```
With `-m` we need to run `python -m baseline_name.main`:
```suggestion
8. Ensure that your baseline runs with default argument by running `poetry run python -m <your_baseline_name>.main`. Then, follow the instructions provided in the `README.md` and detail the steps to follow in `Environment Setup` and in `Running the Experiments`.
```

danieljanes commented on baselines/baseline_template/EXTENDED_README.md:
```diff
@@ -0,0 +1,52 @@
+
+More critically however, is adding your ML framework of choice to the list of dependencies. 
+
+TODO
```
Should this be removed before merging?

jafermarq replied:
yes. I'll add some info soon

jafermarq replied:
now updated!

danieljanes commented on baselines/baseline_template/README.md:
```diff
@@ -0,0 +1,81 @@
+
+> :warning: Please follow the instructions in `EXTENDED_README.md`
+
+> :warning: Please follow the instructions carefully. First copy this directory and rename it with something meaningful about your baseline. Then add the code into the Python scripts provided, and edit all the fields in the readme that start with a :warning: symbols (and remove the comments as you start adding the information). You can see the [FedProx-Mnist baseline](~https://github.com/adap/flo ...
```
```suggestion
> :warning: Please follow the instructions carefully. First copy this directory and rename it with something meaningful about your baseline. Then add the code into the Python scripts provided, and edit all the fields in the readme that start with a :warning: symbols (and remove the comments as you start adding the information). You can see the [FedProx-MNIST baseline](~https://github.com/adap/flower/tree/fedprox_mnist_refresh/baselines/flwr_baselines/publications/fedprox_mnist~) as an example of a baseline that followed this guide.
```

danieljanes commented on baselines/baseline_template/README.md:
```diff
@@ -0,0 +1,81 @@
+
+****Labels:**** :warning: *_add a list of single-word (maybe two-words) terms that can be used to categorise this baseline. For example: `image classification`, `heterogeneous clients`, `personalisation`, `communication efficiency`. Please use between 4 and 10 labels._*
+
+****Paper:**** :warning: *_add the URL of the paper._*
```
Should we ask people to link to the arXiv page instead of the PDF?

danieljanes commented on baselines/baseline_template/README.md:
```diff
@@ -0,0 +1,81 @@
+
+## Environment Setup
+
+:warning: _The Python environmet for all baselines should follow these guidelines in the `EXTENDED_README`. Specify the steps to create and activate your environment. If there are any external system-wide requirements please include instructions for them too. These instructions should be comprenhensive enough so anyone can run them (if non standard, describe them step-by-step)._
```
```suggestion
:warning: _The Python environment for all baselines should follow these guidelines in the `EXTENDED_README`. Specify the steps to create and activate your environment. If there are any external system-wide requirements, please include instructions for them too. These instructions should be comprehensive enough so anyone can run them (if non-standard, describe them step-by-step)._
```

danieljanes commented on baselines/baseline_template/baseline_template/client.py:
```diff
@@ -0,0 +1,5 @@
+##################################################################################
+# Please use this script to define your client class and a function to construct #
+# such clients. Most of the code here is expected to be defined inside a class   #
+# that inherits from `flwr.client.NumPyClient`.                                  #
```
Should we also mention `flwr.client.Client`?

danieljanes commented on baselines/baseline_template/baseline_template/conf/base.yaml:
```diff
@@ -0,0 +1,17 @@
+  _target_: #points to your startegy (either custom or exiting in Flower)
+  # rest of strategy config
+
+clinet:
+  # client config
```
```suggestion
client:
  # client config

```

danieljanes commented on baselines/baseline_template/baseline_template/conf/base.yaml:
```diff
@@ -0,0 +1,17 @@
+dataset:
+  # dataset config
+
+moodel:
```
```suggestion
model:
```

danieljanes commented on baselines/baseline_template/pyproject.toml:
```diff
@@ -0,0 +1,87 @@
+    "Programming Language :: Python",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3 :: Only",
+    "Programming Language :: Python :: 3.7",
```
Remove 3.7 (we'll remove support soon), but add 3.10 and 3.11

danieljanes commented on baselines/dev/create_baselie.sh:
```diff
@@ -0,0 +1,24 @@
+
+# adjusting package name in pyproject.toml
+echo "!!! Your directory for your baseline '$name' is almost ready"
+echo "!!! Now go to $name/pyproject.toml and ensure that 'name' under '[tool.poetry]' is set to: name = \"$name\""
```
```suggestion
echo "!!! Now go to $name/pyproject.toml and ensure that 'name' under '[tool.poetry]' is set to: name = \"$name\""

```

danieljanes commented on baselines/dev/create_baselie.sh:
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
```
Let's rename the file using a `-` instead of an `_`, and also `baseline` instead of `baselie` 🤣

jafermarq commented:
@danieljanes I have made the changes you suggested. Also added some extra information so contributors have a bit more guidance on how to add dependencies via `poetry add`. Finally, I added the `metadata` section at the top of the `README.md` in the template. This metadata auto generates a table with the minimal information about a baseline: currently `paper title`, `URL`, `labels`. Should we add more? maybe `datasets` ?

danieljanes commented on baselines/baseline_template/pyproject.toml:
```diff
@@ -0,0 +1,88 @@
+ignore_missing_imports = true
+strict = false
+plugins = "numpy.typing.mypy_plugin"
+
+
```
```suggestion

```

danieljanes commented on baselines/baseline_template/pyproject.toml:
```diff
@@ -0,0 +1,88 @@
+]
+
+[tool.poetry.dependencies]
+python = ">=3.8.15, <=3.11.0"  # don't change this
```
Is `<=3.11.0` correct or do we actually want `<3.12`? We'd want to allow `3.11.1`, right?

jafermarq replied:
i really don't know what the best practice is for this. But allowing below `3.12` sounds good to me.

danieljanes commented on baselines/baseline_template/baseline_template/dataset_preparation.py:
```diff
@@ -0,0 +1,43 @@
+# please uncomment the lines below and tell us in the README.md (see the "Running  #
+# the Experiment" block) that this file should be executed first.                  #
+####################################################################################
+
+
+
+
+
+
+
```
```suggestion


```

danieljanes commented on baselines/baseline_template/baseline_template/conf/base.yaml:
```diff
@@ -0,0 +1,17 @@
+  # model config
+
+strategy:
+  _target_: #points to your startegy (either custom or exiting in Flower)
```
```suggestion
  _target_: # points to your strategy (either custom or exiting in Flower)
```

danieljanes commented on baselines/baseline_template/pyproject.toml:
```diff
@@ -0,0 +1,88 @@
+
+[tool.pylint.messages_control]
+disable = "bad-continuation,duplicate-code,too-few-public-methods,useless-import-alias"
+signature-mutators="hydra.main.main"
```
Let's finish all files with a newline

jafermarq replied:
got it.

danieljanes commented on baselines/baseline_template/pyproject.toml:
```diff
@@ -0,0 +1,88 @@
+
+[tool.pylint.messages_control]
+disable = "bad-continuation,duplicate-code,too-few-public-methods,useless-import-alias"
+signature-mutators="hydra.main.main"
```
```suggestion
signature-mutators="hydra.main.main"

```

danieljanes commented on baselines/baseline_template/baseline_template/models.py:
```diff
@@ -0,0 +1,8 @@
+
```
```suggestion
```

danieljanes commented on baselines/baseline_template/baseline_template/models.py:
```diff
@@ -0,0 +1,8 @@
+##################################################################################
+# Please use this script to define your models, and training and eval            #
+# functions. If your model is 100% off-the-shelf (e.g. directly from torchvision #
+# without requiring modifications) you might be better of instantiating your     #
```
```suggestion
# without requiring modifications) you might be better off instantiating your     #
```

danieljanes commented on baselines/baseline_template/baseline_template/server.py:
```diff
@@ -0,0 +1,6 @@
+##############################################################################
+# Please use this script to define the function called by your strategy in   #
```
Let's use "module" in all the comments. A script would imply that this `.py` can be run, whereas a module implies that it's a `.py` files that just contains Python functions and classes.

```suggestion
# Please use this module to define the function called by your strategy in   #
```

danieljanes commented on baselines/baseline_template/baseline_template/conf/base.yaml:
```diff
@@ -0,0 +1,17 @@
+  # rest of strategy config
+
+client:
+  # client config
```
```suggestion
  # client config

```

danieljanes commented on baselines/baseline_template/baseline_template/main.py:
```diff
@@ -0,0 +1,59 @@
+    # Hydra will generate for you a directory each time you run the code. You
+    # can retrieve the path to that directory with this:
+    # save_path = HydraConfig.get().runtime.output_dir
+    # 
```
```suggestion

```

danieljanes commented on baselines/baseline_template/baseline_template/models.py:
```diff
@@ -0,0 +1,7 @@
+##################################################################################
+# Please use this module to define your models, and training and eval            #
+# functions. If your model is 100% off-the-shelf (e.g. directly from torchvision #
+# without requiring modifications) you might be better off instantiating your     #
```
```suggestion
# without requiring modifications) you might be better off instantiating your    #
```

danieljanes commented on baselines/baseline_template/pyproject.toml:
```diff
@@ -0,0 +1,86 @@
+]
+
+[tool.poetry.dependencies]
+python = ">=3.8.15, <=3.12.0"  # don't change this
```
Not like this?
```suggestion
python = ">=3.8.15, <3.12.0"  # don't change this
```

danieljanes commented on baselines/baseline_template/pyproject.toml:
```diff
@@ -0,0 +1,86 @@
+
+[tool.poetry.dependencies]
+python = ">=3.8.15, <=3.12.0"  # don't change this
+flwr = "1.3.0" # don't change this
```
```suggestion
flwr = "1.4.0" # don't change this
```

danieljanes commented on baselines/baseline_template/pyproject.toml:
```diff
@@ -0,0 +1,86 @@
+include_trailing_comma = true
+force_grid_wrap = 0
+use_parentheses = true
+known_first_party = ["flwr_baselines"]
```
Is this needed? If so, does it need to match the `name` attribute on line 6?

jafermarq replied:
i don't know if this is needed. I copied the structure used in the `pyproject.toml` previous versions of the baselines. ([isort docs](https://pycqa.github.io/isort/docs/configuration/options.html#known-first-party))

danieljanes replied:
Let's remove it then

danieljanes commented on baselines/baseline_template/pyproject.toml:
```diff
@@ -0,0 +1,86 @@
+
+[tool.black]
+line-length = 88
+target-version = ["py38", "py39"]
```
```suggestion
target-version = ["py38", "py39", "py310", "py311"]
```

danieljanes commented on baselines/baseline_template/pyproject.toml:
```diff
@@ -0,0 +1,86 @@
+include_trailing_comma = true
+force_grid_wrap = 0
+use_parentheses = true
+known_first_party = ["flwr_baselines"]
```
```suggestion
```
